use a finalizer instead of GC to clean up templateinstances

bparees opened the pull request:
bug https://bugzilla.redhat.com/show_bug.cgi?id=1568607

bparees commented:
@deads2k @liggitt ptal

bparees commented:
/retest

bparees commented:
/retest

bparees commented:
/retest

deads2k commented on pkg/cmd/server/bootstrappolicy/controller_policy.go:
```diff
@@ -172,6 +172,19 @@ func init() {
+	addControllerRole(rbac.ClusterRole{
+		ObjectMeta: metav1.ObjectMeta{Name: saRolePrefix + InfraTemplateInstanceFinalizerControllerServiceAccountName},
+		Rules: []rbac.PolicyRule{
+			rbac.NewRule("update").Groups(templateGroup).Resources("templateinstances/status").RuleOrDie(),
```
reading top to bottom, but I expected you to need broad delete powers.

bparees replied:
it's got project admin just like the template instance controller.  would you still expect something else?

deads2k commented on pkg/template/apis/template/constants.go:
```diff
@@ -39,4 +39,12 @@ const (
+
+	// TemplateInstanceOwner is a label applied to all objects created from a template instance
+	// which contains the uid of the template instance.
+	TemplateInstanceOwner = "template.openshift.io/template-instance-owner"
```
again reading top to bottom, but this screams: "migration problem" to me.

bparees replied:
templateinstances created by old versions won't have this finalizer, and resources created by those templateinstances will have ownerrefs on them, which means they'll get handled by normal GC and the finalizer should ignore them.

templateinstances created by this version will have the finalizer and the objects created by the templateinstance will have the label and no ownerref, so GC will ignore them and finalization will cover it.

At least that is my theory.  Other than "i was creating a template instance but it's objects weren't created yet and then i upgraded my controller" do you see a hole?  I don't think i can convince myself to care about that scenario.

deads2k replied:
I thought this was solving a cross namespace problem for you.  Don't you need to have the cluster update to be in a state where it's solved?

bparees replied:
not really, no.  anyone who created cross-namespace content already (which i suspect is no one) has already potentially experienced that content getting spontaneously erased by the GC (since as far as the GC is concerned, that content has no valid ownerref already).

We might do a knowledge base about cleaning up that potential content, but again i think cross-namespace object creation is a really rare use case.  In particular because it would have to be coupled with either TSB based instantiation, or direct templateinstance creation, which no one knows how to do except us.

deads2k commented on pkg/template/controller/templateinstance_controller.go:
```diff
@@ -72,18 +74,19 @@ type TemplateInstanceController struct {
-		queue:            workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "openshift_template_instance_controller"),
-		readinessLimiter: workqueue.NewItemFastSlowRateLimiter(5*time.Second, 20*time.Second, 200),
-		clock:            clock.RealClock{},
+		restmapper:        restutil.DefaultMultiRESTMapper(),
```
this restmapper looks compiled in.  Why are you passing one in and using this?

bparees replied:
may have to talk in person about this one.  I didn't change this code, I only added the dynamicRestMapper.  (and per the discussion in the other PR, we might not need this mapper at all anymore but right now the checkReadiness codepath is using it)

deads2k commented on pkg/template/controller/templateinstance_finalizer.go:
```diff
@@ -0,0 +1,279 @@
+
+	// FIXME: Remove then cient when the build configs are able to report the
+	//				status of the last build.
+	buildClient buildclient.Interface
```
top to bottom, but this looks weird.  Why?

bparees replied:
why does it look weird or why is it here? :)

looks like more cut+paste sloppiness on my part, will remove.

deads2k commented on pkg/template/controller/templateinstance_finalizer.go:
```diff
@@ -0,0 +1,279 @@
+	//				status of the last build.
+	buildClient buildclient.Interface
+
+	kc kclientsetinternal.Interface
```
top to bottom, but this looks weird.  Why?

bparees replied:
i must have pulled this from my sample finalizer and didn't use it here, will remove.

deads2k commented on pkg/template/controller/templateinstance_finalizer.go:
```diff
@@ -0,0 +1,279 @@
+// using its own service account, first verifying that the requester also has
+// permissions to instantiate.
+type TemplateInstanceFinalizerController struct {
+	restmapper        meta.RESTMapper
```
why have this restmapper too?

bparees replied:
cut+paste, will remove.

deads2k commented on pkg/template/controller/templateinstance_finalizer.go:
```diff
@@ -0,0 +1,279 @@
+// permissions to instantiate.
+type TemplateInstanceFinalizerController struct {
+	restmapper        meta.RESTMapper
+	dynamicRestMapper *discovery.DeferredDiscoveryRESTMapper
```
why do you need to know it's a particular kind of restmapper?

bparees replied:
OCD :)

will change to interface.

deads2k commented on pkg/template/controller/templateinstance_finalizer.go:
```diff
@@ -0,0 +1,279 @@
+	restmapper        meta.RESTMapper
+	dynamicRestMapper *discovery.DeferredDiscoveryRESTMapper
+	client            dynamic.DynamicInterface
+	config            *rest.Config
```
why are there two of these?

bparees replied:
cut+paste, will remove.

deads2k commented on pkg/template/controller/templateinstance_finalizer.go:
```diff
@@ -0,0 +1,279 @@
+	kc kclientsetinternal.Interface
+
+	lister   templatelister.TemplateInstanceLister
+	informer cache.SharedIndexInformer
```
this is unusual, why are you tracking it?

bparees replied:
you mean vs adding my handler and throwing away the ref?  the controller Run waits for the cache to be synced.

bparees replied:
ie 	`if !cache.WaitForCacheSync(stopCh, c.informer.HasSynced) {`

deads2k replied:
We only keep a ref to the synced functions

bparees replied:
updated in `david 2`

deads2k commented on pkg/template/controller/templateinstance_finalizer.go:
```diff
@@ -0,0 +1,279 @@
+		return err
+	}
+
+	needsFinalizing := false
```
check deletiontimestamp again.  If some cleared a finalizer, then manually deleted, then recreated with the same name, you'd end up here again and you should skip action.

deads2k commented on pkg/template/controller/templateinstance_finalizer.go:
```diff
@@ -0,0 +1,279 @@
+		}
+
+		namespaced := mapping.Scope.Name() == meta.RESTScopeNameNamespace
+		apiResource := &metav1.APIResource{Name: mapping.Resource, Group: gv.Group, Version: gv.Version, Kind: o.Ref.Kind, Namespaced: namespaced}
```
why are you constructing this?

bparees replied:
dead code from an earlier attempt at this. will remove.

deads2k commented on test/testdata/bootstrappolicy/bootstrap_cluster_role_bindings.yaml:
```diff
@@ -821,6 +821,36 @@ items:
+    annotations:
+      rbac.authorization.kubernetes.io/autoupdate: "true"
+    creationTimestamp: null
+    name: admin
```
ah, I see.  Makes sense now.

deads2k replied:
> ah, I see. Makes sense now.

Name doesn't look good through.  `system:openshift:controller:<something>`

bparees replied:
i used the same name that templateinstancecontroller was using...should that be changed too?  (will reconciliation handle it?)

bparees replied:
(actually this is what `NewClusterBinding` gives back by default as the name..uses the rolename)

deads2k commented:
I would encourage a unit test. I take it you're stuck on https://github.com/kubernetes/kubernetes/pull/63253 ?

bparees commented:
I wasn't aware of being stuck on anything other than your review comments.

I've hopefully addressed all your comments either in response form or in the new commit ("david 1").

Unit test for what aspect?

bparees commented:
(and of course this PR sits on top of the CRD PR, so i need to get you happy w/ that PR before this one can merge)

bparees commented on test/integration/authorization_test.go:
```diff
@@ -622,7 +622,7 @@ func TestAuthorizationResourceAccessReview(t *testing.T) {
 				Groups:          sets.NewString(),
 				Namespace:       "mallet-project",
-				EvaluationError: `[clusterrole.rbac.authorization.k8s.io "admin" not found, clusterrole.rbac.authorization.k8s.io "admin" not found]`,
+				EvaluationError: `[clusterrole.rbac.authorization.k8s.io "admin" not found, clusterrole.rbac.authorization.k8s.io "admin" not found, clusterrole.rbac.authorization.k8s.io "admin" not found]`,
```
@enj i don't really understand why my changes in this PR would necessitate this change... can you enlighten me on what is going on here? (in general it seems weird that we'd return a string containing the same error message repeated N times w/ no additional context)

liggitt replied:
checking who can do a thing involves checking all applicable role bindings and the roles they refer to. any bindings that reference an unresolveable role result in a lookup error. all the lookup errors are aggregated here: https://github.com/openshift/origin/blob/master/vendor/k8s.io/kubernetes/plugin/pkg/auth/authorizer/rbac/subject_locator.go#L79

I'm not sure if we should dedupe the aggregated messages or add more info to distinguish which binding each lookup error came from (which could leak info).

In any case, you added a new clusterrolebinding which in this test becomes invalid (since the test intentionally deletes the admin role). This test might be better served creating a single-use role and a couple referencing rolebindings.

bparees replied:
thanks @liggitt, makes sense

bparees commented:
/retest

deads2k commented on pkg/template/controller/templateinstance_finalizer.go:
```diff
@@ -0,0 +1,261 @@
+		templateClient:    templateClient,
+		lister:            informer.Lister(),
+		informerSynced:    informer.Informer().HasSynced,
+		queue:             workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "openshift_template_instance_controller"),
```
name

bparees replied:
fixing

deads2k commented on pkg/template/controller/templateinstance_finalizer.go:
```diff
@@ -0,0 +1,261 @@
+	})
+
+	c.jsonConfig = rest.CopyConfig(config)
+	c.jsonConfig.ContentConfig = dynamic.ContentConfig()
```
don't need to set this

bparees replied:
removing

deads2k commented on pkg/template/controller/templateinstance_finalizer.go:
```diff
@@ -0,0 +1,261 @@
+		},
+	})
+
+	c.jsonConfig = rest.CopyConfig(config)
```
don't need to copy this

bparees replied:
removing

deads2k commented on pkg/template/controller/templateinstance_finalizer.go:
```diff
@@ -0,0 +1,261 @@
+
+	needsFinalizing := false
+	for _, v := range templateInstance.Finalizers {
+		if v == templateapi.TemplateInstanceFinalizer && templateInstance.DeletionTimestamp != nil {
```
move the deletiontimestamp check out of the for loop.  Just check above and return.

bparees replied:
fixing

deads2k commented on pkg/template/controller/templateinstance_finalizer.go:
```diff
@@ -0,0 +1,261 @@
+	for _, o := range templateInstance.Status.Objects {
+		glog.V(5).Infof("attempting to delete object: %#v", o)
+
+		gv, err := schema.ParseGroupVersion(o.Ref.APIVersion)
```
group and version should be different fields in your API.  API already baked?

deads2k replied:
> also i don't think we have a choice unless we move away from using an ObjectReference ?

Yeah, I'd avoid an object reference . We've defined specific types in most new code.

bparees replied:
yeah so that ship sailed when the templateinstance api was introduced.

deads2k commented on pkg/template/controller/templateinstance_finalizer.go:
```diff
@@ -0,0 +1,261 @@
+		}
+
+		var resourceClient dynamic.DynamicResourceInterface
+		if namespaced {
```
this gets less ugly after the rebase.

bparees replied:
yeah i saw, i'll clean it up at that point i guess.

deads2k commented on pkg/template/controller/templateinstance_finalizer.go:
```diff
@@ -0,0 +1,261 @@
+
+	errs := []error{}
+	background := metav1.DeletePropagationBackground
+	deleteOpts := &metav1.DeleteOptions{PropagationPolicy: &background}
```
why background?

bparees replied:
didn't see a reason to hold up finalization waiting for GC to do its job on any child resources, but if that's not the right approach i'm happy to change it.

deads2k replied:
> didn't see a reason to hold up finalization waiting for GC to do its job on any child resources, but if that's not the right approach i'm happy to change it.

I think you want foreground.  It still sets a finalizer and you get an early return, but the object stays around until the descendants are removed.

deads2k commented on pkg/template/controller/templateinstance_finalizer.go:
```diff
@@ -0,0 +1,261 @@
+func (c *TemplateInstanceFinalizerController) Run(workers int, stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	defer c.queue.ShutDown()
+
```
glog.Info for starting, ending, and cache sync please.  They pop out once each and are super useful.

bparees replied:
adding

bparees commented on pkg/template/controller/templateinstance_finalizer.go:
```diff
@@ -0,0 +1,261 @@
+}
+
+// enqueueAfter adds a TemplateInstance to c.queue after a duration.
+func (c *TemplateInstanceFinalizerController) enqueueAfter(templateInstance *templateapi.TemplateInstance, duration time.Duration) {
```
was unused, removing.

bparees commented:
comments addressed or fixed in `david 4`

bparees commented:
/retest

bparees commented:
/retest

bparees commented:
@deads2k putting this back on your radar

deads2k commented on pkg/template/controller/templateinstance_finalizer.go:
```diff
@@ -0,0 +1,255 @@
+			Group: gv.Group,
+			Kind:  o.Ref.Kind,
+		}
+		mapping, err := c.dynamicRestMapper.RESTMapping(gk, gv.Version)
```
make a comment that doing this means that if an api resource type is removed, this finalizer will fail here and hang.

bparees replied:
ok

bparees commented:
comments addressed in `david 4`

deads2k commented on pkg/template/controller/templateinstance_finalizer.go:
```diff
@@ -0,0 +1,259 @@
+		namespaced := mapping.Scope.Name() == meta.RESTScopeNameNamespace
+		namespace := ""
+		if namespaced {
+			namespace = o.Ref.Namespace
```
just to confirm.  A normal user cannot touch this ref, right?

bparees replied:
it's part of the templateinstance status so they should not be able to, no.

deads2k commented on pkg/template/controller/templateinstance_finalizer.go:
```diff
@@ -0,0 +1,259 @@
+		}
+	}
+	if len(errs) > 0 {
+		return kerrs.NewAggregate(errs)
```
@liggitt do we allow status to be updated on objects being finalized?

liggitt replied:
afaik, all updates (both to spec and to status) are allowed

deads2k commented on pkg/template/controller/templateinstance_finalizer.go:
```diff
@@ -0,0 +1,259 @@
+		}
+	}
+	if len(errs) > 0 {
+		return kerrs.NewAggregate(errs)
```
@bparees I strongly recommend emitting an event either here or on the general error handling for debugging later.

bparees replied:
ok

deads2k commented on pkg/template/controller/templateinstance_finalizer.go:
```diff
@@ -0,0 +1,259 @@
+		return
+	}
+
+	glog.V(2).Infof("Starting TemplateInstanceFinalizer controller")
```
I recommend putting this at the info level so you know that you've sync'd.

bparees replied:
will do

deads2k commented on test/extended/testdata/templates/templateservicebroker_bind.yaml:
```diff
@@ -61,18 +61,6 @@ objects:
           to:
             kind: Service
             name: service
-      - apiVersion: v1
```
why the change?

bparees replied:
it's really reverting a change (which was made in the other templateinstance PR).  the legacy/grouped type handling is checked in templateinstance_objectkinds.yaml, so i wanted to put this back the way i found it for simplicity.

deads2k commented on test/testdata/bootstrappolicy/bootstrap_cluster_role_bindings.yaml:
```diff
@@ -812,7 +812,7 @@ items:
       rbac.authorization.kubernetes.io/autoupdate: "true"
     creationTimestamp: null
-    name: admin
+    name: template-instance-controller:admin
```
`system:openshift:controller:template-instance-controller:admin`  "don't muck with this: it's for openshift: for a controller: this controller: bound to that role"

deads2k commented on test/testdata/bootstrappolicy/bootstrap_cluster_role_bindings.yaml:
```diff
@@ -821,6 +821,36 @@ items:
+    annotations:
+      rbac.authorization.kubernetes.io/autoupdate: "true"
+    creationTimestamp: null
+    name: template-instance-finalizer-controller:admin
```
system:openshift:controller:etc

deads2k commented:
minor comments.  lgtm otherwise

bparees commented:
comments addressed in `david 5`

deads2k commented:
lgtm, please squash before merge

After the rebase we'll have a fake dynamic client.  I recommend writing a unit test.

bparees commented:
squashed.

looks like the switch to foreground may be messing w/ one of the tests, will investigate if it fails again.
